File: java/array_search/HashSearch.java
import java.util.HashMap;
import java.util.Map;

public class HashSearch {

    public static int hashSearch(int[] arr, int target) {
        Map<Integer, Integer> indexByValue = new HashMap<>();

        for (int i = 0; i < arr.length; i++) {
            indexByValue.putIfAbsent(arr[i], i);
        }

        return indexByValue.getOrDefault(target, -1);
    }
}

File: cpp/array_search/hash_search.cpp
#include <unordered_map>
#include <vector>

using namespace std;

int hash_search(const vector<int>& arr, int target) {
    unordered_map<int, int> table;

    for (int i = 0; i < static_cast<int>(arr.size()); i++) {
        if (table.find(arr[i]) == table.end()) {
            table[arr[i]] = i;
        }
    }

    auto it = table.find(target);
    return it == table.end() ? -1 : it->second;
}
